File: Services/ErrorLogger.vb
Imports System.IO
Imports System.Text

''' <summary>エラーログ出力</summary>
Public NotInheritable Class ErrorLogger
    Private Sub New()
    End Sub

    Public Shared Sub Write(ex As Exception, Optional context As String = Nothing)
        Try
            Dim logPath = System.IO.Path.Combine(AppContext.BaseDirectory, "error.log")
            Dim sb As New StringBuilder()
            sb.AppendLine($"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] {If(context, "Error")}")
            sb.AppendLine(ex.ToString())
            sb.AppendLine(New String("-"c, 60))
            File.AppendAllText(logPath, sb.ToString(), Encoding.UTF8)
        Catch
            ' ログ自体の失敗は握りつぶす
        End Try
    End Sub

    Public Shared Sub Write(message As String)
        Try
            Dim logPath = System.IO.Path.Combine(AppContext.BaseDirectory, "error.log")
            File.AppendAllText(logPath, $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] {message}{Environment.NewLine}", Encoding.UTF8)
        Catch
        End Try
    End Sub
End Class

File: Services/ErrorLogger.cs
using System.Text;

namespace KeyAutomator.Services;

public static class ErrorLogger
{
    /// <summary>直近に書き込めたログファイルのフルパス。書けなければ null。</summary>
    public static string? LastWrittenPath { get; private set; }

    public static void Write(Exception ex, string? context = null)
    {
        var sb = new StringBuilder();
        sb.AppendLine($"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] {context ?? "Error"}");
        sb.AppendLine(ex.ToString());
        sb.AppendLine(new string('-', 60));
        WriteRaw(sb.ToString());
    }

    public static void Write(string message)
    {
        WriteRaw($"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] {message}{Environment.NewLine}");
    }

    private static void WriteRaw(string text)
    {
        // 主経路（DataDirectory）→ だめなら Temp
        var candidates = new[]
        {
            AppPaths.ErrorLogPath,
            Path.Combine(Path.GetTempPath(), "KeyAutomator-error.log")
        };

        foreach (var path in candidates.Distinct(StringComparer.OrdinalIgnoreCase))
        {
            try
            {
                var dir = Path.GetDirectoryName(path);
                if (!string.IsNullOrEmpty(dir))
                    Directory.CreateDirectory(dir);
                File.AppendAllText(path, text, Encoding.UTF8);
                LastWrittenPath = path;
                return;
            }
            catch
            {
                // try next
            }
        }
    }
}
